tests/test_allow_none.py

```diff
@@ -0,0 +1,120 @@
+
+Permission is hereby granted, free of charge, to any person obtaining a
+copy of this software and associated documentation files (the
+"Software"), to deal in the Software without restriction, including
+without limitation the rights to use, copy, modify, merge, publish,
+distribute, sublicense, and/or sell copies of the Software, and to
+permit persons to whom the Software is furnished to do so, subject to
+the conditions stated in the LICENSE file in the project root for
+details.
+
+The above copyright notice and this permission notice shall be included
+in all copies or substantial portions of the Software.
+"""
+
+
+import unittest
+
+import configsuite
+from configsuite import MetaKeys as MK
+from .data import computers
+
+
+class TestAllowNone(unittest.TestCase):
+    def check_f(self, key, explicit):
+        schema = computers.build_schema()
+        config = computers.build_config()
+
+        schema[MK.Content][key][MK.AllowNone] = True
+
+        if explicit:
+            config[key] = None
+        else:
+            config.pop(key)
+        config_suite = configsuite.ConfigSuite(config, schema)
+
+        self.assertTrue(config_suite.valid)
+        self.assertIsNone(config_suite.snapshot.__dict__[key])
+
+    def test_basictype_explicit_none(self):
+        self.check_f("OS", explicit=True)
+
+    def test_basictype_implicit_none(self):
+        self.check_f("OS", explicit=False)
+
+    def test_dict_explicit_none(self):
+        self.check_f("software", explicit=True)
+
+    def test_dict_implicit_none(self):
+        self.check_f("software", explicit=False)
+
+    def test_named_dict_explicit_none(self):
+        self.check_f("components", explicit=True)
+
+    def test_named_dict_implicit_none(self):
+        self.check_f("components", explicit=False)
+
+    def test_list_explicit_none(self):
+        self.check_f("externals", explicit=True)
+
+    def test_list_implicit_none(self):
+        self.check_f("externals", explicit=False)
+
+    def test_not_required_implies_must_allow_none(self):
+        schema = computers.build_schema()
+        config = computers.build_config()
+
+        schema[MK.Content]["OS"][MK.Required] = False
+        schema[MK.Content]["OS"][MK.AllowNone] = False
+
+        with self.assertRaises(ValueError):
+            configsuite.ConfigSuite(config, schema)
+
+
+class TestNotAllowNone(unittest.TestCase):
+    def check_f(self, key, error_contains, explicit):
+        schema = computers.build_schema()
+        config = computers.build_config()
+
+        if explicit:
+            config[key] = None
+        else:
+            config.pop(key)
+        config_suite = configsuite.ConfigSuite(config, schema)
+
+        self.assertFalse(config_suite.valid)
+        for err in config_suite.errors:
+            self.assertTrue(
+                isinstance(err, configsuite.validation_errors.InvalidTypeError)
+            )
+            self.assertTrue(err.msg.startswith(error_contains))
```

Comment:
Just a minor thing. `error_contains` suggest that the value passed here is checked as `assert error_contains in err.msg`, but we do a `startswith` instead

Reply:
Changed the asserts of errors